server/subscribe.go

```diff
@@ -0,0 +1,455 @@
+			return nil, err
+		}
+
+		subscription, ok := subs.Channel.ById[subscriptionId]
+		if !ok {
+			return nil, errors.New("could not find subscription")
+		}
+
+		subs.Channel.remove(&subscription)
+
+		modifiedBytes, marshalErr := json.Marshal(&subs)
+		if marshalErr != nil {
+			return nil, marshalErr
+		}
+
+		return modifiedBytes, nil
+	})
+}
+
+func (p *Plugin) addChannelSubscription(newSubscription *ChannelSubscription) error {
+	return p.atomicModify(JIRA_SUBSCRIPTIONS_KEY, func(initialBytes []byte) ([]byte, error) {
+		subs, err := SubscriptionsFromJson(initialBytes)
+		if err != nil {
+			return nil, err
+		}
+
+		newSubscription.Id = model.NewId()
+		subs.Channel.add(newSubscription)
+
+		modifiedBytes, marshalErr := json.Marshal(&subs)
+		if marshalErr != nil {
+			return nil, marshalErr
+		}
+
+		return modifiedBytes, nil
+	})
+}
+
+func (p *Plugin) editChannelSubscription(modifiedSubscription *ChannelSubscription) error {
+	return p.atomicModify(JIRA_SUBSCRIPTIONS_KEY, func(initialBytes []byte) ([]byte, error) {
+		subs, err := SubscriptionsFromJson(initialBytes)
+		if err != nil {
+			return nil, err
+		}
+
+		oldSub, ok := subs.Channel.ById[modifiedSubscription.Id]
+		if !ok {
+			return nil, errors.New("Existing subscription does not exist.")
+		}
+		subs.Channel.remove(&oldSub)
+		subs.Channel.add(modifiedSubscription)
+
+		modifiedBytes, marshalErr := json.Marshal(&subs)
+		if marshalErr != nil {
+			return nil, marshalErr
+		}
+
+		return modifiedBytes, nil
+	})
+}
+
+func (p *Plugin) atomicModify(key string, modify func(initialValue []byte) ([]byte, error)) error {
+	readModify := func() ([]byte, []byte, error) {
+		initialBytes, appErr := p.API.KVGet(key)
+		if appErr != nil {
+			return nil, nil, errors.Wrap(appErr, "unable to read inital value")
+		}
+
+		modifiedBytes, err := modify(initialBytes)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "modification error")
+		}
+
+		return initialBytes, modifiedBytes, nil
+	}
+
+	success := false
+	for !success {
+		initialBytes, newValue, err := readModify()
+		if err != nil {
+			return err
+		}
+
+		var setError *model.AppError
+		success, setError = p.API.KVCompareAndSet(key, initialBytes, newValue)
+		if setError != nil {
+			return errors.Wrap(setError, "problem writing value")
+		}
+
+	}
+
+	return nil
+}
+
+func httpSubscribeWebhook(p *Plugin, w http.ResponseWriter, r *http.Request) (int, error) {
+	if r.Method != http.MethodPost {
+		return http.StatusMethodNotAllowed,
+			fmt.Errorf("Request: " + r.Method + " is not allowed, must be POST")
+	}
+
+	cfg := p.getConfig()
+	if cfg.Secret == "" || cfg.UserName == "" {
+		return http.StatusForbidden, fmt.Errorf("JIRA plugin not configured correctly; must provide Secret and UserName")
+	}
+
+	if subtle.ConstantTimeCompare([]byte(r.URL.Query().Get("secret")), []byte(cfg.Secret)) != 1 {
+		return http.StatusForbidden, fmt.Errorf("Request URL: secret did not match")
+	}
+
+	parsed, err := parse(r.Body, nil)
```

Comment:
`defer r.Body.Close()` should go above the parse call

Reply:
Added.